File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */

 /*
 1. l = 0, r = last height
    area = 0
 2. while (l < r) {
    currentArea = (r - l) * Min(height[l], height[r])
    area = Max(area, currentArea)
    if (l < r) {
        l++
    } else {
        r--
    }
 }
 3. return area
 */
var maxArea = function(height) {
    var l = 0
    var r = height.length - 1
    var area = 0

    while(l < r) {
        var currentArea = (r - l) * Math.min(height[l], height[r])
        area = Math.max(area, currentArea)

        if (height[l] < height[r]) {
            l++
        } else {
            r--
        }
    }
    return area
};

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        area = 0

        while (l < r):
            base = r - l
            h = min(height[l], height[r])
            tempArea = base * h
            area = max(area, tempArea)

            if (height[l] <= height[r]):
                l += 1
            elif (height[r] < height[l]):
                r -= 1

        return area